### ISSUE TITLE
MultiGauge metrics stops working from spring boot release 2.2.7 and up (tested up to including 2.3.2)

### ISSUE THREAD

**Issue by bertiljeppsson** [micrometer-metrics/micrometer#2250]
( Moved from springboot https://github.com/spring-projects/spring-boot/issues/23166 )

When we upgraded from spring boot release 2.2.5 to 2.3.2 we noticed out `MultiGauge` metrics stopped working as expected. After getting some initial values for a set of tags, they never changed so we ended up with flat dashboard graphs even though we could see from our application logs that the gauges were updated multiple times to new values.

Our application is running java 11 (openjdk).

I have tested below versions so it seems the issue have been introduced with the 2.2.7 release:
2.2.5 -> works
2.2.6 -> works
2.2.7 -> fails
2.3.0 -> fails
2.3.1 -> fails
2.3.2 -> fails

Micrometer was updated to version [1.3.8](https://github.com/spring-projects/spring-boot/issues/21185) in spring boot version 2.2.7, which could indicate that the issue was introduced with that version.

The problem we observe is the following.
We have several `MultiGauge` meters which we update periodically using the content of a Map with `AtomicInteger` where the Metadata type contains a unique set of tags we wish to report the metric with.

```
private static final String TAG1 = "tag1";
private static final String TAG2 = "tag2";
private static final String TAG3 = "tag3";

private Map<Metadata, AtomicInteger> pendingChanges;
private MultiGauge gauge;

...
// Called periodically and the content of the pendingChanges map is slightly different on each iteration.
this.gauge.register(pendingChanges.entrySet().stream().map(t -> MultiGauge.Row.of(
                Tags.of(
                        TAG1, t.getKey().tag1,
                        TAG2, t.getKey().tag2,
                        TAG3, t.getKey().tag3),
                t.getValue())).collect(Collectors.toList()), true);
```

Furthermore, we use `micrometer-registry-prometheus` so that our metrics scraper can extract metrics from the /actuator/prometheus endpoint.
However, what we observe is that the overwrite:true flag of above `MultiGauge.register(...)` no longer appears to work as expected.
Once we have reported a unique set of tags with e.g. tag1="a", tag2="b", tag3="c" and value Atomicinteger(37), it appears we are no longer able to overwrite this value. It stays 37 for the lifetime of the application even though we make multiple attempts to overwrite it for the same set of tags.

We have not been able to reproduce the problem in our unit tests, so it appears to work when we update the gauges multiple times and after verify the values from the global metric registry using:
`Metrics.globalRegistry.find("metricName")`

The affected 'frozen' metric values are seen on both /actuator/prometheus as well as /actuator/metrics/metricName, so it does not appear to be specific to prometheus, but rather something internal in the in-memory meter registry for some reason won't be updated to the new values.

**Comment by arvs7** [comment]
We had a similar issue when upgrading springboot. The overwrite of MultiGauge metrics failed as we were setting custom tags using `management.metrics.tags.*` 

Adding these tags to `MultiGauge.Row` fixed the issue for us.

**Comment by izeye** [comment]
@bertiljeppsson @arvs7 Thanks for the report.

This seems to be a side effect of #2025. I created #2283 to try to resolve this.

**Comment by shakuzen** [comment]
The fix for this should be available in the latest snapshots for the upcoming 1.1.19, 1.3.15, 1.5.6 releases. Feel free to test with snapshots and let us know if there is any issue remaining.

**Comment by bertiljeppsson** [comment]
Hi @shakuzen,

Sorry for the delay in following up on this. 
Today we tried out the fix and upgraded to spring 2.3.6.RELEASE which includes micrometer.io 1.5.7 which should include the fix.

However the problem unfortunately appears to persist. The first time we increment the MultiGauge counter, it gets the value 1.0 set with all the attached tags. But when try to increment it further or reset it to zero additional times, it never changes, so we are stuck with a frozen 1.0 value that appears to persist forever.

`$> curl -sk https://localhost:1234/actuator/prometheus | grep multi_gauge_counter
# HELP multi_gauge_counter
# TYPE multi_gauge_counter gauge
multi_gauge_counter{tag1="a",tag2="b",tag3="c",tag4="d",tag5="e",tag6="f",} 1.0` 

Can we re-open this issue? I don't seem to be able to do that, but if needed I can open a new bug and copy the content once more.

**Comment by shakuzen** [comment]
I think it'll be better to make a new issue referencing this issue. If possible, could you provide a unit test that reproduces the problem so we can make sure we're fixing the right issue for you?
